cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,160 @@
+import json
+import logging
+import re
+import subprocess
+from importlib.resources import files
+from typing import Optional, Union
+
+import pydantic
+
+from cachi2.core.errors import PackageRejected, UnexpectedFormat
+from cachi2.core.rooted_path import PathOutsideRoot, RootedPath
+
+log = logging.getLogger(__name__)
+
+GEMFILE = "Gemfile"
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class _GemMetadata(pydantic.BaseModel):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(_GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this gem dependency."""
+        return NotImplemented
+
+
+class GitDependency(_GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        if not url.startswith("https://"):
+            raise ValueError("GIT dependencies must use HTTPS protocol")
+
+        return url
+
+    @pydantic.field_validator("revision")
+    def validate_git_revision(cls, revision: str) -> str:
+        """Validate that the git revision is a valid git ref."""
+        git_ref_format = re.compile(r"^[a-fA-F0-9]{40}$")
+        if not git_ref_format.match(revision):
```

Comment:
No need to do it this way, you can just use `re.match` directly as it'll compile the pattern implicitly. If you want to leverage the benefit of compiling patterns upfront, then you need to cache it somewhere so that it'll be used for every single git dependency model instantiation, but admittedly that only becomes useful in decreasing overhead once you need to process a good number of strings (several hundreds, maybe even thousands), which I don't expect to be the case here, so just plain `re.match` is going to be sufficient for us.